src/lib.rs

```diff
@@ -1,7 +1,754 @@
 #![no_std]
 
-#[cfg(not(feature = "binary-vendor"))]
-mod contract;
+use gear_lib::non_fungible_token::token::TokenMetadata;
+use gstd::{errors::Result as GstdResult, exec, msg, prelude::*, ActorId, MessageId};
+use hashbrown::{HashMap, HashSet};
+use hint::unreachable_unchecked;
+use supply_chain_io::*;
+use tx_manager::{TransactionGuard, TransactionManager};
+
+mod tx_manager;
+mod utils;
 
 #[cfg(feature = "binary-vendor")]
 include!(concat!(env!("OUT_DIR"), "/wasm_binary.rs"));
+
+fn get_mut_item(
+    items: &mut HashMap<ItemId, Item>,
+    item_id: ItemId,
+    expected_item_state: ItemState,
+) -> Result<&mut Item, Error> {
+    let item = items.get_mut(&item_id).ok_or(Error::ItemNotFound)?;
+
+    if item.info.state != expected_item_state {
+        return Err(Error::UnexpectedItemState);
+    }
+
+    Ok(item)
+}
+
+fn role_to_set_item_dr(role: Role) -> fn(&mut Item, ActorId) {
+    const FNS: [fn(&mut Item, ActorId); 2] = [Item::set_distributor, Item::set_retailer];
+
+    FNS[role as usize - 1]
+}
+
+type IsPdr = fn(&Item, ActorId) -> Result<(), Error>;
+
+fn role_to_is_pdr(role: Role) -> IsPdr {
+    const FNS: [IsPdr; 3] = [Item::is_producer, Item::is_distributor, Item::is_retailer];
+
+    FNS[role as usize]
+}
+
+fn role_to_item_pdr(role: Role) -> fn(&Item) -> ActorId {
+    const FNS: [fn(&Item) -> ActorId; 3] = [Item::producer, Item::distributor, Item::retailer];
+
+    FNS[role as usize]
+}
+
+#[derive(Default)]
+struct Item {
+    info: ItemInfo,
+    shipping_time: u64,
+}
+
+impl Item {
+    fn set_retailer(&mut self, retailer: ActorId) {
+        self.info.retailer = retailer
+    }
+
+    fn set_distributor(&mut self, distributor: ActorId) {
+        self.info.distributor = distributor
+    }
+
+    fn set_state_and_get_event(&mut self, item_id: ItemId, item_state: ItemState) -> Event {
+        self.info.state = item_state;
+
+        Event {
+            item_id,
+            item_state,
+        }
+    }
+
+    fn is_pdr(pdr: ActorId, actor_id: ActorId) -> Result<(), Error> {
+        if pdr != actor_id {
+            Err(Error::AccessRestricted)
+        } else {
+            Ok(())
+        }
```

Comment:
Just as an option, not sure it is more readable though 😄 
```suggestion
        (pdr == actor_id).then_some(()).ok_or(Error::AccessRestricted)
```

Reply:
I prefer the current version more. A conversion to `Option` may be confusing for a reader.